dot/core/consensus.go

```diff
@@ -0,0 +1,177 @@
+	log.Debug(
+		"[core] initializing BABE session...",
+		"epoch", s.epochNumber,
+	)
+
+	// AuthorityData comes from NextEpochDescriptor within the ConsensusDigest
+	// of the block Digest, which is included in the first block of each epoch
+	authData := s.bs.AuthorityData()
+	if len(authData) == 0 {
+		return nil, fmt.Errorf("authority data not set")
+	}
+
+	newBlocks := make(chan types.Block)
+	s.blkRec = newBlocks
+
+	epochDone := make(chan struct{})
+	s.epochDone = epochDone
+
+	babeKill := make(chan struct{})
+	s.babeKill = babeKill
+
+	keys := s.keys.Sr25519Keypairs()
+
+	// get best slot to determine next start slot
+	bestHash := s.blockState.BestBlockHash()
+	bestSlot, err := s.blockState.GetSlotForBlock(bestHash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get slot for latest block: %s", err)
+	}
+
+	// BABE session configuration
+	bsConfig := &babe.SessionConfig{
+		Keypair:          keys[0].(*sr25519.Keypair),
+		Runtime:          s.rt,
+		NewBlocks:        newBlocks, // becomes block send channel in BABE session
+		BlockState:       s.blockState,
+		StorageState:     s.storageState,
+		TransactionQueue: s.transactionQueue,
+		AuthData:         authData,
+		Done:             epochDone,
+		Kill:             babeKill,
+		StartSlot:        bestSlot + 1,
+		SyncLock:         s.syncLock,
+	}
+
+	// create new BABE session
+	bs, err := babe.NewSession(bsConfig)
+	if err != nil {
+		log.Error("[core] failed to initialize BABE session", "error", err)
+		return nil, err
+	}
+
+	log.Debug(
+		"[core] BABE session initialized",
+		"epoch", s.epochNumber,
+	)
+
+	return bs, nil
+}
+
+// handleConsensusDigest checks if the provided header is the block header for
+// the first block of the current epoch, finds and decodes the consensus digest
+// item from the block digest, and then sets the epoch data for the next epoch
+func (s *Service) handleConsensusDigest(header *types.Header) (err error) {
+
+	// check if block header digest items exist
+	if header.Digest == nil || len(header.Digest) == 0 {
+		return fmt.Errorf("header digest not set")
+	}
+
+	// declare digest item
+	var item types.DigestItem
+
+	// decode each digest item and check its type
+	for _, digest := range header.Digest {
+		item, err = types.DecodeDigestItem(digest)
+		if err != nil {
+			return err
+		}
+
+		// check if digest item is consensus digest type
+		if item.Type() == types.ConsensusDigestType {
+
+			// cast decoded consensus digest item to conensus digest type
+			consensusDigest := item.(*types.ConsensusDigest)
+
+			// initialize epoch data interface for next epoch
+			nextEpochData := new(babe.NextEpochDescriptor)
+
+			// decode consensus digest data for next epoch
+			err = nextEpochData.Decode(consensusDigest.Data)
+			if err != nil {
+				return err
+			}
+
+			// set epcoh data for next epoch
+			err = s.bs.SetEpochData(nextEpochData)
+			if err != nil {
+				return err
+			}
+
+			// set first block in current epoch
+			s.firstBlock = header.Number
```

Comment:
not sure about this, this assumes that the block we have received is honest/valid, but it might notactually be the first block and this will overwrite it

Reply:
right, I added the check first block method when finalizing an epoch and starting a new babe session and meant to add the check here as well. I was going to assume the first block to claim it was the first block is the first block and error otherwise. Not sure how else to address.

Reply:
okay, sure that seems fair. yeah just make sure this can't get overwritten by subsequent blocks

Reply:
added `verifyFirstBlock` if `ConsensusDigestType`